sycl/include/CL/sycl/detail/device_impl.hpp

```diff
@@ -21,44 +20,141 @@ namespace sycl {
-
-  // Returns underlying native device object (if any) w/o reference count
-  // modification. Caller must ensure the returned object lives on stack only.
-  // It can also be safely passed to the underlying native runtime API.
-  // Warning. Returned reference will be invalid if device_impl was destroyed.
-  //
-  virtual RT::PiDevice &getHandleRef() = 0;
-  virtual const RT::PiDevice &getHandleRef() const = 0;
-
-  virtual bool is_host() const = 0;
-
-  virtual bool is_cpu() const = 0;
-
-  virtual bool is_gpu() const = 0;
-
-  virtual bool is_accelerator() const = 0;
-
-  virtual platform get_platform() const = 0;
+  /// Constructs a SYCL device instance as a host device.
+  device_impl();
+  /// Constructs a SYCL device instance using the provided
+  /// PI device instance.
+  explicit device_impl(RT::PiDevice Device);
+
+  ~device_impl();
+
+  /// Get instance of OpenCL device
+  ///
+  /// @return a valid cl_device_id instance in accordance with the requirements
+  /// described in 4.3.1.
+  cl_device_id get() const;
+
+  /// Get reference to PI device
+  ///
+  /// For host device an exception is thrown
+  ///
+  /// @return non-constant reference to PI device
+  RT::PiDevice &getHandleRef() {
+    if (MIsHostDevice)
+      throw invalid_object_error("This instance of device is a host instance");
+
+    return MDevice;
+  }
+
+  /// Get constant reference to PI device
+  ///
+  /// For host device an exception is thrown
+  ///
+  /// @return constant reference to PI device
+  const RT::PiDevice &getHandleRef() const {
+    if (MIsHostDevice)
+      throw invalid_object_error("This instance of device is a host instance");
+
+    return MDevice;
+  }
+
+  /// Check if device is a host device
+  ///
+  /// @return true if SYCL device is a host device
+  bool is_host() const { return MIsHostDevice; }
+
+  /// Check if device is a CPU device
+  ///
+  /// @return true if SYCL device is a CPU device
+  bool is_cpu() const { return (MType == PI_DEVICE_TYPE_CPU); }
+
+  /// Check if device is a GPU device
+  ///
+  /// @return true if SYCL device is a GPU device
+  bool is_gpu() const { return (MType == PI_DEVICE_TYPE_GPU); }
+
+  /// Check if device is an accelerator device
+  ///
+  /// @return true if SYCL device is an accelerator device
+  bool is_accelerator() const { return (MType == PI_DEVICE_TYPE_ACC); }
+
+  /// Get associated SYCL platform
+  ///
+  /// If this SYCL device is an OpenCL device then the SYCL platform
+  /// must encapsulate the OpenCL cl_plaform_id associated with the
+  /// underlying OpenCL cl_device_id of this SYCL device. If this SYCL device
+  /// is a host device then the SYCL platform must be a host platform.
+  /// The value returned must be equal to that returned
+  /// by get_info<info::device::platform>().
+  ///
+  /// @return The associated SYCL platform.
+  platform get_platform() const;
+
+  /// Check SYCL extension support by device
+  ///
+  /// @extension_name is a name of queried extension.
```

Comment:
Will it generate correct doxygen doc? If I understand correctly, this should be something like `@param ExtensionName`